### DISCUSSION TITLE
Command line API compared to restic

### DISCUSSION THREAD

**Discussion by nils-werner** [rustic-rs/rustic#609] [Q&A] [answered]
Hi,

great work, I'm really happy to see restic ported to rust!

I'm the author of [crestic](https://github.com/nils-werner/crestic), a configuration wrapper for restic, and I'm interested in adding support for rustic.

My question is: how does the command line API of rustic compare to restic? Are they drop in replaceable?

**Comment by aawsome** [comment, marked as answer]
The CLI options are similar to restic, but not 100% compatible. Some options have different names (e.g. `--host` in restic became `--filter-host` in rustic) and there are of course many additional options in rustic.

However, I fear that a wrapper like crestic is not needed for rustic as it already has built-in config file support.
Using a config file like `myprofile.toml`:
```
[repository]
repository = "/tmp/backup"
password = "my password"

[[backup.sources]]
source = "/home/"
git-ignore = true
```
allows you to call `rustic -P myprofile backup` and it will do the job. See the `examples` dir for more sophisticated config files.

**Comment by nils-werner** [comment]
crestic can actually do more than just load a profile: [split profile names](https://nils-werner.github.io/crestic/config/commands.html#split-keys)

**Comment by aawsome** [comment]
@nils-werner Thanks for the hint! With rustic you can do similar things by splitting your storage config and source config in multiple config files and then using both, e.g. `rustic -P home -P local backup`.
Or you could even use the config file inheritance creating a new config `home@local.toml` which only consists of 
```
[global]
use-profile = ["home", "local"]
```
and then run `rustic -P home@local backup`. 
Also if there are other missing features about configuration, I'm sure it will be added to rustics's native config file support.

That said, I'd like to absolutely encourage you to add rustic to crestic if you think the crestic users would benefit from it. `rustic -help` or `rustic help <COMMAND>` should give you all needed information about the CLI API.

**Comment by nils-werner** [comment]
Ah that's neat, too!

As a matter of fact, crestic can already talk to rustic, because it's simply setting command line options, and doesn't care what option names are. All you have to do is set the `CRESTIC_EXECUTABLE=rustic` environment variable.

I was just asking about the option names to see if the same identical config files could be used for both restic and rustic.

**Comment by simonsan** [comment]
@nils-werner Would be neat, if you could collect features from your `crestic` POV that `rustic` with its configuration files is still missing and opening an issue for it, so we can discuss how/if to include that. 

We could also work together on writing migration guides e.g. `migrating from restic` and `migrating from crestic configuration`.

**Reply by nils-werner** [reply]
Not sure if I have enough material for individual issues, but one thing I noticed is that there are three kinds of options that I had to represent in the INI syntax

```
--key "value"
--empty ""        # empty string
--flag            # options without values
```

I had to configure the Python `ConfigParser` to enable the syntax

```
[home.backup]
key: value
empty:
flag
```

and I am not sure if this is representable in TOML

**Reply by nils-werner** [reply]
I've created a PR #739

**Comment by aawsome** [comment]
> ```
> [home.backup]
> key: value
> empty:
> flag
> ```
> 
> and I am not sure if this is representable in TOML

I think that these three options would read like this in TOML:
```toml
[home.backup]
key = "value" 
empty = ""
flag = true
```
And, specifically withing a rustic configuration you would call `rustic -P home backup` using a `home.toml` containing something like
```toml
[[backup.sources]]
source = "/home"
key = "value" 
empty = ""
flag = true
```
